aesd-char-driver/main.c

```diff
@@ -132,42 +147,190 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
-            if (free_buffptr != NULL)
-            {
-                kfree(free_buffptr);
-                free_buffptr = NULL;
-            }
-            dev->entry.buffptr = NULL;
-            dev->entry.size = 0;
-        }
+    		free_buffptr = aesd_circular_buffer_add_entry(&dev->buffer, &dev->entry);
+	        // free previously allocated entry if any
+	        if (free_buffptr != NULL)
+	        {
+	            kfree(free_buffptr);
+	            free_buffptr = NULL;
+	        }
+	        
+	        dev->entry.buffptr = NULL;
+	        dev->entry.size = 0;
+	}
     }while(0);
     
-    
     // release lock
     mutex_unlock(&dev->lock);
     
+    return retval;
+}
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    int entry_index = 0;
+    loff_t file_offset = 0;
+    loff_t total_size = 0;
+    struct aesd_dev *dev = NULL;
+    struct aesd_buffer_entry *entry = NULL;
     
+    PDEBUG("aesd_llseek()");
+
+    dev = filp->private_data;
+	
+    // acquire lock
+    if (0 != mutex_lock_interruptible(&dev->lock))
+    {
+        PDEBUG("ERROR: mutex_lock_interruptible acquiring lock");
+        return -ERESTARTSYS;
+    }
+
+    // to get the total size
+    AESD_CIRCULAR_BUFFER_FOREACH(entry,&aesd_device.buffer,entry_index)
+    {
+        total_size += entry->size;
+    }
+	
+    // release lock
+    mutex_unlock(&dev->lock);
+
+    file_offset = fixed_size_llseek(filp, off, whence, total_size);
     
-    return retval;
+    return file_offset;
 }
+
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+	long retval = 0;
+	int i;
+	int entry_length;
+	struct aesd_dev *dev = filp->private_data;
+	struct aesd_buffer_entry *entry = NULL;
+	
+	PDEBUG("aesd_adjust_file_offset()");
+
+	// acquire lock
+    	if (mutex_lock_interruptible(&dev->lock) != 0)
+	{
+		PDEBUG("mutex_lock_interruptible() acquiring lock error");
+		return -ERESTARTSYS;
+	}
+	
+	do
+	{
+		PDEBUG("aesd_adjust_file_offset() start");
+		// check if write_cmd exceeds max writes supported
+		if (write_cmd > AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+		{		
+			PDEBUG("invalid");
+			retval = -EINVAL;
+			break; // release lock and exit
+	    	}
+	    	
+	    	// check if write_cmd exceeds no. of entries present
+	    	AESD_CIRCULAR_BUFFER_FOREACH(entry,&aesd_device.buffer,entry_length);
+		if (write_cmd > entry_length)
+		{
+			PDEBUG("invalid");
+			retval = -EINVAL;
+			break; // release lock and exit
+	    	}
+	    	
+		// check if offset exceeds size of entry
+		if (write_cmd_offset >= dev->buffer.entry[write_cmd].size)
+		{
+			PDEBUG("invalid");
+			retval = -EINVAL;
+			break; // release lock and exit
+	    	}
+	    	
+	    	// adjust the file offset position
+	    	for (i=0; i<write_cmd; i++)
+		{
+			filp->f_pos += dev->buffer.entry[i].size;
+		}
+		filp->f_pos += write_cmd_offset;
+		
+		PDEBUG("aesd_adjust_file_offset() completed");
+		
+	}while(0);
```

Comment:
@ameyflash what's the purpose of the do while here?

Reply:
That is to exit correctly by unlocking the mutex instead of using goto statements I use do...while(0).

Reply:
I see.. I'd probably refactor into a function call instead and make it clear that the caller must hold lock when calling.

Reply:
Got it. Thanks!